libraries/AP_RangeFinder/AP_RangeFinder_GYUS42v2.cpp

```diff
@@ -0,0 +1,72 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include "AP_RangeFinder_GYUS42v2.h"
+#include <ctype.h>
+
+extern const AP_HAL::HAL& hal;
+
+bool AP_RangeFinder_GYUS42v2::find_signature_in_buffer(uint8_t start)
+{
+    for (uint8_t i=start; i<buffer_used; i++) {
+        if (buffer[i] == 0x5A) {
+            memmove(&buffer[0], &buffer[i], buffer_used-i);
+            buffer_used -= i;
+            return true;
+        }
+    }
+    // header byte not in buffer
+    buffer_used = 0;
+    return false;
+}
+
+// get_reading - read a value from the sensor
+bool AP_RangeFinder_GYUS42v2::get_reading(uint16_t &reading_cm)
+{
+    if (uart == nullptr) {
+        return false;
+    }
+
+    const uint8_t num_read = uart->read(&buffer[buffer_used], ARRAY_SIZE(buffer)-buffer_used);
+    buffer_used += num_read;
+
+    if (buffer_used == 0) {
+        return false;
+    }
+
+    if (buffer[0] != 0x5A &&
+        !find_signature_in_buffer(1)) {
```

Comment:
It's not a blocker but an alternative implementation would have been to throw away bytes until we see the header byte.  The LeddarVu8 driver has this kind of state machine (so does the Benewake serial driver). https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_RangeFinder/AP_RangeFinder_LeddarVu8.cpp#L131

Reply:
One advantage in doing it the way presented here is that if your checksum doesn't match then you can scan the buffer and potentially find a real packet starting in the middle of what you thought was your payload.

If you're running a state machine you usually don't do that - you just throw away what didn't match and start from "want header" again.